packages/4-xdapp-layer/aptos/src/context.ts

```diff
@@ -0,0 +1,319 @@
+import { BigNumber } from 'ethers';
+import { sha3_256 } from 'js-sha3';
+import { arrayify, hexlify, stripZeros, zeroPad } from 'ethers/lib/utils';
+import { AptosClient, CoinClient, Types } from 'aptos';
+import {
+  TokenId,
+  ParsedRelayerMessage,
+  ChainName,
+  ChainId,
+  NATIVE,
+  ParsedMessage,
+  Context,
+  ParsedRelayerPayload,
+  Wormhole,
+  hexToUint8Array,
+  parseTokenTransferPayload,
+  TokenBridgeAbstract,
+  MAINNET_CHAINS,
+  Network,
+} from '@wormhole-foundation/sdk-base';
+import { AptosContracts } from './contracts';
+import {
+  getForeignAssetAptos,
+  getIsTransferCompletedAptos,
+  getTypeFromExternalAddress,
+  isValidAptosType,
+  redeemOnAptos,
+  transferFromAptos,
+} from './utils';
+
+export const APTOS_COIN = '0x1::aptos_coin::AptosCoin';
+
+/**
+ * @category Aptos
+ */
+export class AptosContext extends TokenBridgeAbstract<Types.EntryFunctionPayload> {
+  readonly type = Context.APTOS;
+  readonly contracts: AptosContracts;
+  protected wormhole: Wormhole;
+  readonly aptosClient: AptosClient;
+  readonly coinClient: CoinClient;
+
+  constructor(network: Network, wormholeInstance?: Wormhole) {
+    super();
+    this.wormhole = wormholeInstance || new Wormhole(network, {});
+    const rpc = this.wormhole.conf.rpcs.aptos;
+    if (rpc === undefined) throw new Error('No Aptos rpc configured');
+    this.aptosClient = new AptosClient(rpc);
+    this.coinClient = new CoinClient(this.aptosClient);
+    this.contracts = new AptosContracts(this.wormhole, this.aptosClient);
+  }
+
+  async startTransfer(
+    token: TokenId | typeof NATIVE,
+    amount: bigint,
+    sendingChain: ChainName | ChainId,
+    senderAddress: string,
+    recipientChain: ChainName | ChainId,
+    recipientAddress: string,
+    relayerFee: string = '0',
+  ): Promise<Types.EntryFunctionPayload> {
+    const destContext = this.wormhole.getContext(recipientChain);
+    const recipientChainId = this.wormhole.toChainId(recipientChain);
+
+    let recipientAccount = recipientAddress;
+    // get token account for solana
+    if (recipientChainId === MAINNET_CHAINS.solana) {
+      let tokenId = token;
+      if (token === NATIVE) {
+        tokenId = {
+          address: APTOS_COIN,
+          chain: 'aptos',
+        };
+      }
+      recipientAccount = await this.wormhole.getSolanaRecipientAddress(
```

Comment:
I understand why we need to do some more work for sol, but theres gotta be a better way than including this in each Context obj